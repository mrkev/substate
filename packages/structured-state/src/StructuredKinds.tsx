import { SSet, Structured } from ".";
import { SArray, SSchemaArray } from "./sstate";
import { LinkedPrimitive } from "./state/LinkedPrimitive";
import { Struct } from "./Struct";
import { Struct2 } from "./Struct2";
import { SUnion } from "./sunion";

export type PrimitiveKind = number | string | boolean | null;

// was KnowableObject before
export type StructuredKind =
  | LinkedPrimitive<any>
  | Struct<any>
  | Struct2<any>
  | Structured<any, any>
  | SArray<any>
  | SSchemaArray<any>
  | SSet<any>
  | SUnion<any>;

export function isStructuredKind(val: unknown) {
  return (
    val instanceof LinkedPrimitive ||
    val instanceof Struct ||
    val instanceof Struct2 ||
    val instanceof Structured ||
    val instanceof SArray ||
    val instanceof SSchemaArray ||
    val instanceof SSet ||
    val instanceof SUnion
  );
}
