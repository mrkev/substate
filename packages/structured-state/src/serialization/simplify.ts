import { SArray, SSchemaArray, SSet } from "..";
import { Struct } from "../Struct";
import { Struct2 } from "../Struct2";
import { Structured } from "../Structured";
import {
  isStructuredKind,
  PrimitiveKind,
  StructuredKind,
} from "../StructuredKinds";
import { exhaustive } from "../assertions";
import {
  isSimplified,
  NSimplified,
  Simplified,
  SimplifiedDescriptor,
  SimplifiedRefOf,
} from "./serialization";
import { LinkedPrimitive } from "../state/LinkedPrimitive";
import { CONTAINER_IGNORE_KEYS } from "../state/SubbableContainer";
import { SUnion } from "../sunion";
import { JSONValue } from "../lib/types";
import { OrderedMap } from "../lib/OrderedMap";

export class SimplificationMetadata {
  readonly allIds = new Set<string>();
  readonly allObjs = new Map<string, Simplified>();
  readonly refered = new OrderedMap<string, Simplified>();

  record<T extends Simplified>(obj: StructuredKind, simplified: T): T {
    // console.log("simplifying", obj._id, simplified.$$);
    if (this.allIds.has(obj._id)) {
      throw new Error("Dubplicate " + obj._id);
    }
    this.allIds.add(obj._id);
    this.allObjs.set(obj._id, simplified);
    // return { $$: "ref", _to: obj._id };
    return simplified;
  }

  reference<T extends Simplified>(simplified: T): SimplifiedRefOf<T["$$"]> {
    const present = this.refered.get(simplified._id);
    if (present != null) {
      throw new Error("object already referenced. Not supported yet.");
    }
    this.refered.push(simplified._id, simplified);
    return { $$: "ref", _id: simplified._id, kind: simplified.$$ };
  }
}

export type SimplePackage = {
  nodes: [string, Simplified][];
  // root: NSimplified["ref"];
  simplified: Simplified;
};

export function simplifyAndPackage(state: StructuredKind): SimplePackage {
  const acc = new SimplificationMetadata();
  const simplified = simplifyStructuredKind(state, acc);
  return {
    nodes: Array.from(acc.refered.entries()),
    // todo: find right kind
    // root: { $$: "ref", _id: state._id, kind: "ref" },
    simplified,
  };
}

export function isSimplePackage(json: unknown): json is SimplePackage {
  // TODO: more validation?
  return (
    typeof json === "object" &&
    json != null &&
    !Array.isArray(json) &&
    "nodes" in json &&
    // "root" in json &&
    "simplified" in json &&
    isSimplified(json.simplified)
  );
}

function simplifyPrimitive(
  obj: LinkedPrimitive<any>,
  acc: SimplificationMetadata
): NSimplified["prim"] | SimplifiedRefOf<"prim"> {
  if (obj._container.size > 1) {
    console.warn("multiple containers reference", obj);
  }

  return acc.reference(
    acc.record(obj, {
      $$: "prim",
      _value: obj.get(),
      _id: obj._id,
    })
  );
}

function simplifySimpleArray(
  obj: SArray<any>,
  acc: SimplificationMetadata
): NSimplified["arr-simple"] {
  if (obj._container.size > 1) {
    console.warn("multiple containers reference", obj);
  }
  return acc.record(obj, {
    $$: "arr-simple",
    _value: obj._getRaw().map((x) => simplify(x, acc)),
    _id: obj._id,
  });
}

function simplifySchemaArray(
  obj: SSchemaArray<any>,
  acc: SimplificationMetadata
): NSimplified["arr-schema"] {
  if (obj._container.size > 1) {
    console.warn("multiple containers reference", obj);
  }

  return acc.record(obj, {
    $$: "arr-schema",
    _value: obj._getRaw().map((x) => {
      if (!isStructuredKind(x)) {
        throw new Error("un-knowable found in schema array");
      } else {
        return simplifyStructuredKind(x, acc);
      }
    }),
    _id: obj._id,
  });
}

function simplifyStruct(
  obj: Struct<any>,
  acc: SimplificationMetadata
): NSimplified["struct"] {
  if (obj._container.size > 1) {
    console.warn("multiple containers reference", obj);
  }

  // offer a way to override simplification
  if ("_simplify" in obj && typeof obj._simplify === "function") {
    return {
      $$: "struct",
      _id: obj._id,
      _value: obj._simplify(),
    };
  }

  // const result: Record<any, any> = { _kind: this._kind };
  const result: Record<string, unknown> = {};
  const keys = Object.keys(obj);

  for (const key of keys) {
    if (CONTAINER_IGNORE_KEYS.has(key)) {
      continue;
    }

    const val = (obj as any)[key] as unknown;
    result[key] = simplify(val as any, acc);
  }

  return acc.record(obj, {
    $$: "struct",
    _id: obj._id, // reduntant for typescript
    _value: result,
  });
}

function simplifyStruct2(
  obj: Struct2<any>,
  acc: SimplificationMetadata
): NSimplified["struct2"] {
  if (obj._container.size > 1) {
    console.warn("multiple containers reference", obj);
  }

  return acc.record(obj, {
    $$: "struct2",
    _id: obj._id,
    _value: obj.serialize(),
  });
}

function autoSimplify(
  descriptor: Record<string, StructuredKind | PrimitiveKind>,
  acc: SimplificationMetadata
): SimplifiedDescriptor {
  const serializable: SimplifiedDescriptor = {};
  for (const [key, value] of Object.entries(descriptor)) {
    if (
      typeof value === "string" ||
      typeof value === "number" ||
      typeof value === "boolean" ||
      value == null
    ) {
      serializable[key] = value;
    } else if (typeof value === "function") {
      throw new Error("cant simplify function");
    } else if (value instanceof LinkedPrimitive) {
      serializable[key] = simplifyPrimitive(value, acc);
    } else if (value instanceof SArray) {
      serializable[key] = simplifySimpleArray(value, acc);
    } else if (value instanceof SSchemaArray) {
      serializable[key] = simplifySchemaArray(value, acc);
    } else if (value instanceof Struct) {
      serializable[key] = simplifyStruct(value, acc);
    } else if (value instanceof Struct2) {
      serializable[key] = simplifyStruct2(value, acc);
    } else if (value instanceof Structured) {
      serializable[key] = simplifyStructured(value, acc);
    } else if (value instanceof SSet) {
      serializable[key] = simplifySet(value, acc);
    } else if (value instanceof SUnion) {
      serializable[key] = simplifyUnion(value, acc);
    } else {
      exhaustive(value);
    }
  }

  return serializable;
}

export function simplifyStructured(
  obj: Structured<any, any>,
  acc: SimplificationMetadata
): NSimplified["structured"] {
  if (obj._container.size > 1) {
    console.warn("multiple containers reference", obj);
  }

  return acc.record(obj, {
    $$: "structured",
    _id: obj._id,
    _value: autoSimplify(obj.autoSimplify(), acc),
  });
}

function simplifySet(
  obj: SSet<any>,
  acc: SimplificationMetadata
): NSimplified["set"] {
  if (obj._container.size > 1) {
    console.warn("multiple containers reference", obj);
  }

  return acc.record(obj, {
    $$: "set",
    _schema: obj._schema != null,
    _value: obj.map((x) => {
      return simplify(x, acc);
    }),
    _id: obj._id,
  });
}

function simplifyUnion(
  obj: SUnion<any>,
  acc: SimplificationMetadata
): NSimplified["union"] {
  if (obj._container.size > 1) {
    console.warn("multiple containers reference", obj);
  }

  return acc.record(obj, {
    $$: "union",
    _value: simplifyStructuredKind(obj, acc),
    _id: obj._id,
  });
}

function simplify(
  state: StructuredKind | JSONValue,
  acc: SimplificationMetadata
) {
  if (
    typeof state === "string" ||
    typeof state === "number" ||
    typeof state === "boolean" ||
    state == null
  ) {
    return state;
  } else if (typeof state === "function") {
    throw new Error("cant simplify function");
  } else if (
    state instanceof LinkedPrimitive ||
    state instanceof SArray ||
    state instanceof SSchemaArray ||
    state instanceof Struct ||
    state instanceof Struct2 ||
    state instanceof Structured ||
    state instanceof SSet ||
    state instanceof SUnion
  ) {
    return simplifyStructuredKind(state, acc);
  } else if (typeof state === "object") {
    if (state.constructor !== Object && !Array.isArray(state)) {
      throw new Error("cant simplify non-literal object or array");
    }
    return state;
  } else {
    exhaustive(state);
  }
}

export function simplifyPrimitiveKind(state: string | number | boolean | null) {
  return state;
}

export function simplifyStructuredKind(
  state: StructuredKind,
  acc: SimplificationMetadata
): Simplified {
  if (state instanceof LinkedPrimitive) {
    return simplifyPrimitive(state, acc);
  } else if (state instanceof SArray) {
    return simplifySimpleArray(state, acc);
  } else if (state instanceof SSchemaArray) {
    return simplifySchemaArray(state, acc);
  } else if (state instanceof Struct) {
    return simplifyStruct(state, acc);
  } else if (state instanceof Struct2) {
    return simplifyStruct2(state, acc);
  } else if (state instanceof Structured) {
    return simplifyStructured(state, acc);
  } else if (state instanceof SSet) {
    return simplifySet(state, acc);
  } else if (state instanceof SUnion) {
    return simplifyUnion(state, acc);
  } else {
    exhaustive(state);
  }
}
