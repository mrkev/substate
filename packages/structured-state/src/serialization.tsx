import { initialize } from "./serialization.initialize";
import { simplify } from "./serialization.simplify";
import { SArray, SSchemaArray, SState } from "./sstate";
import { LinkedPrimitive } from "./state/LinkedPrimitive";
import { SSet } from "./state/LinkedSet";
import { Struct } from "./Struct";
import { Struct2 } from "./Struct2";
import { ConstructableStructure, Structured } from "./Structured";
import { StructuredKind } from "./StructuredKinds";

// TODO: map

export type SerializedDescriptor = Record<
  string,
  Serialized | string | number | boolean | null
>;

export type SerializedTypePrimitive<T> = Readonly<{
  $$: "prim";
  _id: string;
  _value: T;
}>;

export type SerializedSimpleArray<T> = Readonly<{
  $$: "arr-simple";
  _id: string;
  _value: readonly T[];
}>;

export type SerializedSimpleSet<T> = Readonly<{
  $$: "arr-simple";
  _id: string;
  _value: readonly T[];
}>;

export type NSimplified = {
  prim: Readonly<{
    $$: "prim";
    _id: string;
    _value: unknown;
  }>;
  "arr-schema": Readonly<{
    $$: "arr-schema";
    _id: string;
    _value: readonly (Serialized | unknown)[];
  }>;
  "arr-simple": Readonly<{
    $$: "arr-simple";
    _id: string;
    _value: readonly unknown[];
  }>;
  struct: Readonly<{
    $$: "struct";
    _id: string;
    _value: {
      [key: string]: Serialized | unknown;
    };
  }>;
  struct2: Readonly<{
    $$: "struct2";
    _id: string;
    _value: readonly unknown[];
  }>;
  structured: Readonly<{
    $$: "structured";
    _id: string;
    _autoValue: SerializedDescriptor;
  }>;
  set: Readonly<{
    $$: "set";
    _id: string;
    _value: readonly (Serialized | unknown)[];
  }>;
  union: Readonly<{
    $$: "union";
    _id: string;
    _value: Serialized;
  }>;
};

export type Serialized = NSimplified[keyof NSimplified];

export type S = {
  string: SerializedTypePrimitive<string>;
  number: SerializedTypePrimitive<number>;
  boolean: SerializedTypePrimitive<boolean>;
  null: SerializedTypePrimitive<null>;
  primitive: NSimplified["prim"];
  arrSchema: NSimplified["arr-schema"];
  arr: NSimplified["arr-simple"];
  struct: NSimplified["struct"];
  struct2: NSimplified["struct2"];
  structured: NSimplified["structured"];
  set: NSimplified["set"];
};

export type ApplySerialization<T extends StructuredKind> =
  T extends LinkedPrimitive<infer U>
    ? SerializedTypePrimitive<U>
    : T extends Struct<any>
    ? NSimplified["struct"]
    : T extends Struct2<any>
    ? NSimplified["struct2"]
    : T extends Structured<any, any>
    ? NSimplified["structured"]
    : T extends SSchemaArray<any>
    ? NSimplified["arr-schema"]
    : T extends SArray<infer U>
    ? SerializedSimpleArray<U>
    : T extends SSet<any>
    ? NSimplified["set"]
    : never;

export type ApplyDeserialization<
  S extends Serialized,
  T extends Struct<any> | Struct2<any> | Structured<any, any> = any
> = S extends NSimplified["prim"]
  ? LinkedPrimitive<any>
  : S extends NSimplified["struct"]
  ? Struct<any>
  : S extends NSimplified["struct2"]
  ? Struct2<any>
  : S extends NSimplified["structured"]
  ? Structured<any, any>
  : S extends NSimplified["arr-simple"]
  ? SArray<any>
  : S extends NSimplified["arr-schema"]
  ? SSchemaArray<T>
  : S extends NSimplified["set"]
  ? SSet<T>
  : never;

export type ObjectDeserialization<
  S extends Serialized,
  T extends Struct<any> | Struct2<any> | Structured<any, any> = any
> = S extends NSimplified["prim"]
  ? LinkedPrimitive<any>
  : S extends NSimplified["struct"]
  ? Struct<any>
  : S extends NSimplified["struct2"]
  ? Struct2<any>
  : S extends NSimplified["structured"]
  ? Structured<any, any>
  : S extends NSimplified["arr-simple"]
  ? T[]
  : S extends NSimplified["arr-schema"]
  ? T[]
  : S extends NSimplified["set"]
  ? Set<T>
  : never;

///////////////////////////////

export function serialize(state: StructuredKind) {
  const allIds = new Set<string>();
  const result = JSON.stringify(simplify(state, { allIds }));
  console.log(allIds);
  return result;
}

///////////////////////////////

/** some initializations require a Schema */
export type Schema =
  | typeof Struct // Struct
  | typeof Struct2 // Struct
  | typeof Structured // Struct
  | (typeof Struct | typeof Struct2 | typeof Structured<any, any>)[]; // SSchemaArray

export type StructSchema =
  | typeof Struct // Struct
  | typeof Struct2 // Struct
  // | typeof Structured; // Struct
  | ConstructableStructure<any>;

export type NeedsSchema =
  | NSimplified["struct"]
  | NSimplified["struct2"]
  | NSimplified["structured"]
  | NSimplified["arr-schema"];

export type NeedsSchemaStruct =
  | Extract<Serialized, { $$: "struct" }>
  | Extract<Serialized, { $$: "struct2" }>
  | Extract<Serialized, { $$: "structured" }>;

export function construct(
  str: string,
  spec:
    | SState<unknown>
    | typeof Struct // Struct
    | (typeof Struct)[] // SArray
) {
  try {
    const json = JSON.parse(str);
    return initialize(json, spec as any);
  } catch (e) {
    console.log("issue with", JSON.parse(str));
    throw e;
  }
}

export function isSeralized(json: unknown): json is Serialized {
  // TODO: more validation?
  return (
    typeof json === "object" &&
    json != null &&
    !Array.isArray(json) &&
    "$$" in json
  );
}

export function isSeralizedStructured(
  json: unknown
): json is NSimplified["structured"] {
  return isSeralized(json) && json.$$ === "structured";
}
