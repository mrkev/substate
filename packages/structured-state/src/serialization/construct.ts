import { Struct } from "../Struct";
import { ConstructableStructure } from "../Structured";
import { StructSchema } from "../StructuredKinds";
import {
  InitializationMetadata,
  initialize,
  initializePrimitive,
  initializeStructured,
} from "./initialize";
import { isSimplePackage, SimplePackage } from "./simplify";

function preInitialize(json: SimplePackage, metadata: InitializationMetadata) {
  console.log(
    "pre-init of nodes",
    json.nodes.map(([_, node]) => `${node.$$}:${node._id}`)
  );
  for (const [id, node] of json.nodes) {
    const _ = initializePrimitive(node as any, metadata);
    console.log("inited", id);
  }
}

export function construct(
  str: string,
  spec:
    | StructSchema
    | StructSchema[]
    // | SState<unknown>
    | typeof Struct // Struct
    | (typeof Struct)[] // SArray
    | null
) {
  try {
    const json = JSON.parse(str);
    if (!isSimplePackage(json)) {
      throw new Error("not a simple package");
    }

    const metadata = new InitializationMetadata(json);
    preInitialize(json, metadata);

    const result = initialize(json.simplified, spec, metadata);
    return result;
  } catch (e) {
    console.log("issue with", JSON.parse(str));
    throw e;
  }
}

export const constructFn = {
  structured<Spec extends ConstructableStructure<any>>(
    str: string,
    spec: Spec
  ) {
    try {
      const json = JSON.parse(str);
      if (!isSimplePackage(json)) {
        throw new Error("not a simple package");
      }

      const metadata = new InitializationMetadata(json);
      preInitialize(json, metadata);

      const simple = json.simplified;
      if (simple.$$ !== "structured") {
        throw new Error(
          `Construction: expected ${"structured"} but got ${json.simplified}`
        );
      }

      initializeStructured(simple, spec, metadata);

      const result = initialize(json.simplified, spec as any, metadata);
      return result;
    } catch (e) {
      console.log("issue with", JSON.parse(str));
      throw e;
    }
  },
};

// function generalConstruct(str: string) {
//   try {
//     const json = JSON.parse(str);
//     if (!isSimplePackage(json)) {
//       throw new Error("not a simple package");
//     }

//     const metadata = new InitializationMetadata(json);
//     const simple = json.simplified;
//     if (simple.$$ !== "structured") {
//       throw new Error(
//         `Construction: expected ${"structured"} but got ${json.simplified}`
//       );
//     }

//     initializeStructured(simple, spec, metadata);

//     const result = initialize(json.simplified, spec as any, metadata);
//     return result;
//   } catch (e) {
//     console.log("issue with", JSON.parse(str));
//     throw e;
//   }
// }
