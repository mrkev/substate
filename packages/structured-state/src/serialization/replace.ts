import {
  assertSPrimitive,
  assertSSchemaArray,
  assertSSet,
  assertSSimpleArray,
  assertStruct,
  assertStruct2,
  assertStructured,
  assertSUnion,
  exhaustive,
} from "../assertions";
import { nullthrows } from "../lib/nullthrows";
import { SArray, SSchemaArray } from "../SArray";
import { LinkedPrimitive } from "../state/LinkedPrimitive";
import { SSet } from "../state/LinkedSet";
import { SubbableContainer } from "../state/SubbableContainer";
import { Struct } from "../Struct";
import { Struct2 } from "../Struct2";
import { Structured } from "../Structured";
import { StructuredKind } from "../StructuredKinds";
import { SUnion } from "../sunion";
import { InitializationMetadata, initialize } from "./initialize";
import { replaceSchemaArray, replaceSimpleArray } from "./replace.array";
import { replaceSSet } from "./replace.set";
import {
  isSimplified,
  NSimplified,
  Simplified,
  SimplifiedRef,
  SimplifiedSet,
  SimplifiedSimpleArray,
  SimplifiedTypePrimitive,
} from "./serialization";
import { isSimplePackage } from "./simplify";

export function replacePackage(json: unknown, obj: StructuredKind): void {
  if (!isSimplePackage(json)) {
    throw new Error("not a simple package");
  }
  const acc = InitializationMetadata.fromPackage(json);
  return replace(json.simplified, obj, acc);
}

export function replace(
  json: Simplified,
  obj: StructuredKind,
  acc: InitializationMetadata
) {
  try {
    switch (json.$$) {
      case "prim": {
        assertSPrimitive(obj);
        return replacePrimitive(json, obj, acc);
      }
      case "arr-schema": {
        assertSSchemaArray(obj);
        return replaceSchemaArray(json, obj, acc);
      }
      case "arr-simple": {
        assertSSimpleArray(obj);
        return replaceSimpleArray(json, obj);
      }
      case "struct": {
        assertStruct(obj);
        return replaceStruct(json, obj);
      }
      case "struct2": {
        assertStruct2(obj);
        return replaceStruct2(json, obj);
      }
      case "structured": {
        assertStructured(obj);
        return replaceStructured(json, obj, acc);
      }
      case "set": {
        assertSSet(obj);
        return replaceSSet(json, obj, acc);
      }
      case "union": {
        assertSUnion(obj);
        return replaceSUnion(json, obj);
      }
      case "ref": {
        const simple = nullthrows(
          acc.knownSimples.get(json._id),
          "ref not found"
        );
        return replace(simple, obj, acc);
        // return replaceRef(json, obj, acc);
      }
      default:
        exhaustive(json, "invalid $$ type");
    }
  } catch (e) {
    console.log("error with replace", json);
    throw e;
  }
}

export function replacePrimitive<T>(
  json: SimplifiedTypePrimitive<T> | SimplifiedRef<"prim">,
  obj: LinkedPrimitive<T>,
  acc: InitializationMetadata
) {
  switch (json.$$) {
    case "prim": {
      obj.replace(json._value);
      return;
    }
    case "ref": {
      const prim = nullthrows(acc.knownSimples.get(json._id), "ref not found");
      // todo, ensure primitive
      obj.replace((prim as any)._value);
      return;
    }

    default:
      exhaustive(json);
  }
}

function replaceStruct(json: NSimplified["struct"], obj: Struct<any>): void {
  // offer a way to override replacement
  if ("_replace" in obj && typeof obj._replace === "function") {
    obj._replace(json._value);
    SubbableContainer._notifyChange(obj, obj);
    return;
  }

  for (const key in json._value) {
    if (key === "$$" || isSimplified(json._value[key])) {
      // TODO: Serialized state gets replaced separately in history (?)
      continue;
    }
    (obj as any)[key] = json._value[key];
  }
  SubbableContainer._notifyChange(obj, obj);
}

function replaceStruct2(json: NSimplified["struct2"], obj: Struct2<any>): void {
  // offer a way to override replacement
  if ("_replace" in obj && typeof obj._replace === "function") {
    obj._replace(json._value);
    SubbableContainer._notifyChange(obj, obj);
    return;
  }

  for (const key in json._value) {
    if (key === "$$" || isSimplified(json._value[key])) {
      // Serialized state gets replaced separately in history (?)
      continue;
    }
    (obj as any)[key] = json._value[key];
  }
  SubbableContainer._notifyChange(obj, obj);
}

export function replaceStructured(
  json: NSimplified["structured"],
  obj: Structured<any, any>,
  acc: InitializationMetadata
) {
  obj.replace(json._value, replaceOfPkg(acc));
  SubbableContainer._notifyChange(obj, obj);
}

export function replaceSUnion(json: NSimplified["union"], obj: SUnion<any>) {
  // todo: Schema. have user deifne a replace function so they can pick the right
  // schema to use when initializing
  // todo: null as any, unimplemented
  const value = initialize(json._value, [], null as any);
  obj.replace(value);
  SubbableContainer._notifyChange(obj, obj);
}

export function replaceOfPkg(
  metadata: InitializationMetadata
): ReplaceFunctions {
  /** `this` on second argument to avoid double-initializing the same elements */
  return {
    string: (
      json: SimplifiedTypePrimitive<string>,
      obj: LinkedPrimitive<string>
    ) => replacePrimitive<string>(json, obj, metadata),
    number: (
      json: SimplifiedTypePrimitive<number>,
      obj: LinkedPrimitive<number>
    ) => replacePrimitive<number>(json, obj, metadata),
    boolean: (
      json: SimplifiedTypePrimitive<boolean>,
      obj: LinkedPrimitive<boolean>
    ) => replacePrimitive<boolean>(json, obj, metadata),
    null: (json: SimplifiedTypePrimitive<null>, obj: LinkedPrimitive<null>) =>
      replacePrimitive<null>(json, obj, metadata),
    primitive: <T>(json: SimplifiedTypePrimitive<T>, obj: LinkedPrimitive<T>) =>
      replacePrimitive(json, obj, metadata),
    schemaArray: <T extends Struct<any> | Struct2<any> | Structured<any, any>>(
      json: NSimplified["arr-schema"],
      arr: SSchemaArray<T>
    ) => replaceSchemaArray(json, arr, metadata),
    array: replaceSimpleArray,
    struct: replaceStruct,
    struct2: replaceStruct2,
    structured: (json: NSimplified["structured"], obj: Structured<any, any>) =>
      replaceStructured(json, obj, metadata),
    set: (json: NSimplified["set"], set: SSet<any>) =>
      replaceSSet(json, set, metadata),
  } as const;
}

export type ReplaceFunctions = Readonly<{
  string: (
    json: SimplifiedTypePrimitive<string>,
    obj: LinkedPrimitive<string>
  ) => void;
  number: (
    json: SimplifiedTypePrimitive<number>,
    obj: LinkedPrimitive<number>
  ) => void;
  boolean: (
    json: SimplifiedTypePrimitive<boolean>,
    obj: LinkedPrimitive<boolean>
  ) => void;
  null: (
    json: SimplifiedTypePrimitive<null>,
    obj: LinkedPrimitive<null>
  ) => void;
  primitive: <T>(
    json: SimplifiedTypePrimitive<T>,
    obj: LinkedPrimitive<T>
  ) => void;
  schemaArray: <T extends Struct<any> | Struct2<any> | Structured<any, any>>(
    json: NSimplified["arr-schema"],
    arr: SSchemaArray<T>
  ) => void;
  array: <T>(json: SimplifiedSimpleArray<T>, arr: SArray<T>) => void;
  struct: (json: NSimplified["struct"], obj: Struct<any>) => void;
  struct2: (json: NSimplified["struct2"], obj: Struct2<any>) => void;
  structured: (
    json: NSimplified["structured"],
    obj: Structured<any, any>
  ) => void;
  set: <T>(json: SimplifiedSet<T>, obj: SSet<T>) => void;
}>;
