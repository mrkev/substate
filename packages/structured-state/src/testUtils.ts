import * as s from "./index";
import { Structured } from "./index";
import { initializeStructured } from "./serialization/initialize";
import { EmptyObject } from "./types";

export class Minimal extends s.Structured<EmptyObject, typeof Minimal> {
  override replace(): void {}
  override autoSimplify(): EmptyObject {
    return {};
  }
  static construct(): Minimal {
    return s.Structured.create(Minimal);
  }

  static withId(id: string) {
    const min = initializeStructured(
      { $$: "structured", _id: id, _value: {} },
      Minimal,
      null
    );
    return min;
  }
}

type ASimple = { num: s.SNumber };
export class Simple extends s.Structured<ASimple, typeof Minimal> {
  constructor(readonly num: s.SNumber) {
    super();
  }

  override autoSimplify() {
    return { num: this.num };
  }

  override replace(
    autoJson: s.JSONOfAuto<ASimple>,
    replace: s.ReplaceFunctions
  ): void {
    replace.number(autoJson.num, this.num);
  }
  static construct(auto: s.JSONOfAuto<ASimple>, init: s.InitFunctions): Simple {
    return Structured.create(Simple, init.number(auto.num));
  }

  static withIds(sid: string, numid: string) {
    const min = initializeStructured(
      {
        $$: "structured",
        _id: sid,
        _value: {
          num: {
            $$: "prim",
            _id: numid,
            _value: 0,
          },
        },
      },
      Simple,
      null
    );
    return min;
  }
}
