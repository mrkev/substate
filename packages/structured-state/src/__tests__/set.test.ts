import { describe, expect, it } from "vitest";
import * as s from "../index";
import { replacePackage } from "../serializaiton.replace";
import { simplifyAndPackage } from "../serialization.simplify";
import { StructSchema } from "../StructuredKinds";
import { Minimal, Simple } from "../testUtils";

const testSet = <T>(initialValue?: Iterable<T>) =>
  s.SSet._create(initialValue, "foobar", undefined);

const testSetOf = <T extends StructSchema>(
  schema: T,
  initialValue?: Iterable<InstanceType<T>>,
  id?: string
) => s.SSet._create(initialValue, id ?? "setid", schema);

// TODO: test replace with non-mins to ensure replace does happen in children

describe("set", () => {
  it("add, hashes up", () => {
    const set = testSet([3]);
    set.add(5);
    expect(set.has(5)).toEqual(true);
    expect(set.size).toEqual(2);
    expect(set._hash).toEqual(1);
  });

  it("replaces", () => {
    const set = testSet([1, 2, 3, 4, 5]);
    const pkg = simplifyAndPackage(set);
    set.add(6);
    set.add(7);
    set.add(8);
    set.add(9);
    set.add(10);
    expect(set.size).toBe(10);

    set.delete(1);
    set.delete(2);
    set.delete(3);

    expect(set.size).toBe(7);
    replacePackage(pkg, set);
    expect(set.size).toBe(5);
    expect(set).toMatchSnapshot();
  });

  it("serializes, constructs", () => {
    const set = testSet([2]);
    const serialized = s.serialize(set);
    expect(serialized).toMatchSnapshot();
    const constructed = s.construct(serialized, null);
    expect(constructed).toMatchSnapshot();
  });
});

describe("setOf", () => {
  it("add, hashes up", () => {
    const set = testSetOf(Minimal, []);
    const min = Minimal.withId("minid");
    set.add(min);
    expect(set.has(min)).toEqual(true);
    expect(set.size).toEqual(1);
    expect(set._hash).toEqual(1);
  });

  it("replaces", () => {
    const [one, two, three] = [
      Minimal.withId("1"),
      Minimal.withId("2"),
      Minimal.withId("3"),
    ];

    const set = testSetOf(Minimal, [
      one,
      two,
      three,
      Minimal.withId("4"),
      Minimal.withId("5"),
    ]);
    const pkg = simplifyAndPackage(set);

    set.add(Minimal.withId("6"));
    set.add(Minimal.withId("7"));
    set.add(Minimal.withId("8"));
    set.add(Minimal.withId("9"));
    set.add(Minimal.withId("10"));
    expect(set.size).toBe(10);

    set.delete(one);
    set.delete(two);
    set.delete(three);

    expect(set.size).toBe(7);

    replacePackage(pkg, set);
    expect(set.size).toBe(5);
    expect(set).toMatchSnapshot();
  });

  it("replace structured", () => {
    const [one] = [Simple.withIds("simple", "num")];

    const set = testSetOf(Simple, [one], "set.rst");
    const pkg = simplifyAndPackage(set);

    expect(one.num.get()).toBe(0);

    one.num.set(1);
    expect(one.num.get()).toBe(1);

    replacePackage(pkg, set);
    expect(one.num.get()).toBe(0);
  });

  it("replace.prepare", () => {
    const [one] = [Minimal.withId("1")];

    const set = testSetOf(Minimal, [one], "set.rst");
    const pkg = simplifyAndPackage(set);

    expect(set.size).toBe(1);
    set.delete(one);
    expect(set.size).toBe(0);

    replacePackage(pkg, set);
    expect(set.size).toBe(1);
  });

  it("serializes, constructs", () => {
    const min = Minimal.withId("minid");
    const set = testSetOf(Minimal, [min]);
    const serialized = s.serialize(set);
    expect(serialized).toMatchSnapshot();
    const constructed = s.construct(serialized, Minimal);
    expect(constructed).toMatchSnapshot();
  });
});
