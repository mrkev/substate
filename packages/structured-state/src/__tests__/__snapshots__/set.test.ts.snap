// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`set > replaces 1`] = `
SSet {
  "_container": Set {},
  "_hash": 16,
  "_id": "foobar",
  "_propagatedTokens": WeakSet {},
  "_schema": null,
  "_set": Set {
    4,
    5,
    1,
    2,
    3,
  },
  "_subscriptors": Set {},
}
`;

exports[`set > serializes, constructs 1`] = `"{"nodes":[],"simplified":{"$$":"set","_schema":false,"_value":[2],"_id":"foobar"}}"`;

exports[`set > serializes, constructs 2`] = `
SSet {
  "_container": Set {},
  "_hash": 0,
  "_id": "foobar",
  "_propagatedTokens": WeakSet {},
  "_schema": null,
  "_set": Set {
    2,
  },
  "_subscriptors": Set {},
}
`;

exports[`setOf > replaces 1`] = `
SSet {
  "_container": Set {},
  "_hash": 18,
  "_id": "setid",
  "_propagatedTokens": WeakSet {},
  "_schema": [Function],
  "_set": Set {
    Minimal {
      "_container": Set {
        [Circular],
      },
      "_hash": 1,
      "_id": "4",
      "_propagatedTokens": WeakSet {},
      "_subscriptors": Set {},
    },
    Minimal {
      "_container": Set {
        [Circular],
      },
      "_hash": 1,
      "_id": "5",
      "_propagatedTokens": WeakSet {},
      "_subscriptors": Set {},
    },
    Minimal {
      "_container": Set {
        [Circular],
      },
      "_hash": 0,
      "_id": "1",
      "_propagatedTokens": WeakSet {},
      "_subscriptors": Set {},
    },
    Minimal {
      "_container": Set {
        [Circular],
      },
      "_hash": 0,
      "_id": "2",
      "_propagatedTokens": WeakSet {},
      "_subscriptors": Set {},
    },
    Minimal {
      "_container": Set {
        [Circular],
      },
      "_hash": 0,
      "_id": "3",
      "_propagatedTokens": WeakSet {},
      "_subscriptors": Set {},
    },
  },
  "_subscriptors": Set {},
}
`;

exports[`setOf > serializes, constructs 1`] = `"{"nodes":[],"simplified":{"$$":"set","_schema":true,"_value":[{"$$":"structured","_id":"minid","_value":{}}],"_id":"setid"}}"`;

exports[`setOf > serializes, constructs 2`] = `
SSet {
  "_container": Set {},
  "_hash": 0,
  "_id": "setid",
  "_propagatedTokens": WeakSet {},
  "_schema": null,
  "_set": Set {
    Minimal {
      "_container": Set {
        [Circular],
      },
      "_hash": 0,
      "_id": "minid",
      "_propagatedTokens": WeakSet {},
      "_subscriptors": Set {},
    },
  },
  "_subscriptors": Set {},
}
`;
